strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        mp = self.scovmat()
+        (A, B, C) = ops.chop_in_blocks_multi(mp, expind)
+        V = A - B @ np.linalg.inv(C + covmat) @ B.transpose(0, 2, 1)
+        self.covs = ops.reassemble_multi(V, expind)
+
+        r = self.smean()
+        (va, vc) = ops.chop_in_blocks_vector_multi(r, expind)
+        va = va + np.einsum("...ij,...j", B @ np.linalg.inv(C + covmat), (vals - vc))
+        self.means = ops.reassemble_vector_multi(va, expind)
+
+        reweights_exp_arg = np.einsum(
+            "...j,...jk,...k", (vals - vc), np.linalg.inv(C + covmat), (vals - vc)
+        )
+        reweights = np.exp(-0.5 * reweights_exp_arg) / (
+            np.sqrt(np.linalg.det(2 * np.pi * (C + covmat)))
+        )
+        self.weights *= reweights
+        self.weights /= np.sum(self.weights)
+
+        self.means = self.means[abs(self.weights) > 0]
+        self.covs = self.covs[abs(self.weights) > 0]
+        self.weights = self.weights[abs(self.weights) > 0]
+
+    def post_select_homodyne(self, n, val, eps=0.0002, phi=0):
+        r"""Simulates a homodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            n (int): mode to be measured
+            val (array): measurement value to post-select
+            eps (int): squeezing of the measurement state
+            phi (float): homodyne angle
+        """
+        if self.active[n] is None:
+            raise ValueError("Cannot apply homodyne measurement, mode does not exist")
+        self.phase_shift(phi, n)
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        indices = [n]
+        vals = np.array([val, 0])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def post_select_heterodyne(self, n, alpha_val):
+        r"""Simulates a heterodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            n (int): mode to be measured
+            alpha_val (array): measurement value to post-select
+        """
+        if self.active[n] is None:
+            raise ValueError("Cannot apply heterodyne measurement, mode does not exist")
+
+        covmat = self.hbar * np.identity(2) / 2
+        indices = [n]
+        vals = np.array([alpha_val.real, alpha_val.imag])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def apply_u(self, U):
+        r"""Transforms the state according to the linear optical unitary that
+        maps a[i] \to U[i, j]^*a[j].
+
+        Args:
+            U (array): linear opical unitary matrix
+        """
+        Us = symp.interferometer(U)
+        self.means = update_means(self.means, Us, self.from_xp)
+        self.covs = update_covs(self.covs, Us, self.from_xp)
+
+    def apply_channel(self, X, Y):
+        r"""Transforms the state according to a deterministic Gaussian CPTP map.
+
+        Args:
+            X (array): matrix for mutltiplicative part of transformation
+            Y (array): matrix for additive part of transformation.
```

Comment:
Would we want to programatically pose some restrictions on `X` and `Y` here?

Reply:
you mean checking that X,Y satisfy some mathematical constraints or some shape constraints?

Reply:
> checking that X,Y satisfy some mathematical constraints

Meant this one. Making sure that they define a `deterministic Gaussian CPTP map`. (Though if this is checked later somewhere in the frontend, then that's also nice)

Reply:
I would probably not check that here.